pkg/kube/krt/debug.go

```diff
@@ -0,0 +1,81 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krt
+
+import (
+	"encoding/json"
+	"sync"
+)
+
+// DebugHandler allows attaching a variety of collections to it and then dumping them
+type DebugHandler struct {
+	debugCollections []DebugCollection
+	mu               sync.RWMutex
+}
+
+func (p *DebugHandler) MarshalJSON() ([]byte, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	return json.Marshal(p.debugCollections)
+}
+
+var GlobalDebugHandler = new(DebugHandler)
+
+type CollectionDump struct {
+	// Map of output key -> output
+	Outputs map[string]any `json:"outputs,omitempty"`
+	// Name of the input collection
+	InputCollection string `json:"inputCollection,omitempty"`
+	// Map of input key -> info
+	Inputs map[string]InputDump `json:"inputs,omitempty"`
+}
+type InputDump struct {
+	Outputs      []string `json:"outputs,omitempty"`
+	Dependencies []string `json:"dependencies,omitempty"`
+}
+type DebugCollection struct {
+	name string
+	dump func() CollectionDump
+}
+
+func (p DebugCollection) MarshalJSON() ([]byte, error) {
+	return json.Marshal(map[string]any{
+		"name":  p.name,
+		"state": p.dump(),
+	})
+}
+
+// maybeRegisterCollectionForDebugging registers the collection in the debugger, if one is enabled
+func maybeRegisterCollectionForDebugging[T any](c Collection[T], handler *DebugHandler) {
+	if handler == nil {
```

Comment:
it could be nice to have an env like `KRT_GLOBAL_DEBUG` that allows enabling debug on all collections without modifying code to apss in `GlobalDebugHandler`